cql3/query_result_consumer.hh

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright (C) 2020 ScyllaDB
+ *
+ * Modified by ScyllaDB
+ */
+
+/*
+ * This file is part of Scylla.
+ *
+ * Scylla is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Scylla is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <seastar/core/sstring.hh>
+
+#include "cql3/result_set.hh"
+#include "transport/event.hh"
+
+namespace cql3 {
+
+class query_result_consumer {
+public:
+    query_result_consumer() = default;
+    query_result_consumer(query_result_consumer&) = delete;
+    virtual void move_to_shard(unsigned shard_id) = 0;
+    virtual void add_warning(const sstring& w) = 0;
+    virtual void set_keyspace(const sstring& keyspace) = 0;
+    virtual void set_result(cql3::result rs) = 0;
+    virtual void set_schema_change(shared_ptr<cql_transport::event::schema_change>& sc) = 0;
+};
+
```

Comment:
I need a better understanding of the motivation. Yes, we avoid creating a result_message, but we now have to pass a callback instead of getting a simple object. We can do that, if there is sufficient justification, but no justification was given.

Reply:
I think you remove an allocation, since result_message is returned as a shared_ptr, but the allocation can also be removed by defining result_message as

```c++
class result_message {
    std::vector<sstring> _warnings;
public:
    class visitor;
    class visitor_base;

    virtual ~result_message() {}

    virtual void accept(visitor&) const = 0;

    void add_warning(sstring w) {
        _warnings.push_back(std::move(w));
    }

    const std::vector<sstring>& warnings() const {
        return _warnings;
    }

    virtual std::optional<unsigned> move_to_shard() const {
        return std::nullopt;
    }
    //
    // Message types:
    //
    std::variant<
            void_message,
            set_keyspace,
            prepared,
            schema_change,
            rows,
            bounce_to_shard,
            > _payload;
};
```

Since you still construct the `rows` object in your patchset (I presume, didn't read it through yet), we end up with the same work.

Reply:
I'm pausing my review until we resolve this point.

Reply:
This is just the first step. Next is to get rid of rows. The intention was to split the change into smaller parts to reduce a risk of conflicts considering the size of the change. Would you like both parts to be in one PR, @avikivity?

Reply:
Get rid of what rows @haaawk?

Reply:
The one Avi mentioned above ;). I believe he meant cql3::result_set::_rows.

Reply:
What's the plan? Before coding, I'd like to understand the direction.

Certainly getting rid of cql3::result_set::_rows is a worthy goal. It is even worth the inversion-of-control (passing a callback to the producer, instead of getting an object from the consumer).

But perhaps we can keep both, by returning a generator instead of a deque.

Reply:
And, for the future, communicate these intents in the cover letter. It's perfectly fine to split such work, but communicating it in advance will make it easier to review and result in less back-and-forth.

Reply:
I think @jul-stas was still preparing the PR. That's why he didn't add any reviewers. It could have been marked cleaner, I guess.

So a generator would be an object we `visit` after it's returned from the method?
I think it would be more complex because we would have to ensure the lifetime of all the parameters and intermediate results is extended as long as generator is not destroyed.

We're still working on the final plan. We will keep you posted.

Reply:
No, a generator is something you call operator() on to generate rows. In general callback syle can be converted to generators. Generators also work well with coroutines - instead of calling `callback(row)`, you call `co_yield row` and the caller calls `row = co_await gen()`.

If you have a callback-style function

```
void do_with_callback(callback c) {
    for (...) {
        callback(row)
     }
}
```

You move the state into a generator object and each call to generator::operator() advances the for loop by one row. Or if using coroutines, you keep the for loop and s/callback/co_yield /.